applications/crb/biotsavart/biotsavart_opt.cpp

```diff
@@ -0,0 +1,166 @@
+    nloptoptions.add(backend_options("backend-l2"));
+    nloptoptions.add(bdf_options("ThermoElectricCRB"));
+    nloptoptions.add(makeOptions());
+
+    Environment env( _argc=argc, _argv=argv,
+                     _desc=nloptoptions,
+                     _about=about(_name="biotsavart_nlopt",
+                                  _author="Feel++ Consortium",
+                                  _email="feelpp-devel@feelpp.org"));
+
+    // List of NLOPT algorithm
+    const boost::unordered_map< const std::string, ::nlopt::algorithm >& authAlgo = boost::assign::map_list_of
+        ("LN_NEWUOA", ::nlopt::LN_NEWUOA )
+        ("LN_COBYLA", ::nlopt::LN_COBYLA )
+        ("LN_BOBYQA", ::nlopt::LN_BOBYQA )
+        ("LD_LBFGS", ::nlopt::LD_LBFGS )
+        ("LD_MMA", ::nlopt::LD_MMA )
+        ("LD_SLSQP", ::nlopt::LD_SLSQP );
+
+    BiotSavartCRB<Thermoelectric> BS = BiotSavartCRB<Thermoelectric>();
+    BS.offline();
+    int N = BS.nbParameters();
+
+    auto salgo = soption("nlopt.algo");
+    Feel::cout << "NLOP algorithm: " << salgo << "\n";
+    auto algo = authAlgo.at(salgo);
+    opt::OptimizationNonLinear opt( algo, N );
+
+    // lower and upper bounds
+    auto muMin = BS.parameterSpace()->min();
+    auto muMax = BS.parameterSpace()->max();
+
+    std::vector<double> lb(muMin.data(), muMin.data()+muMin.size());
+    std::vector<double> ub(muMax.data(), muMax.data()+muMax.size());
+
+    opt.set_lower_bounds(lb);
+    opt.set_upper_bounds(ub);
+
+    // stopping criteria
+    opt.set_maxeval( ioption("nlopt.maxeval") );
+    opt.set_xtol_rel( doption("nlopt.xtol_rel") );
+    opt.set_ftol_rel( doption("nlopt.ftol_rel") );
+    opt.set_xtol_abs( doption("nlopt.xtol_abs") );
+    opt.set_ftol_abs( doption("nlopt.ftol_abs") );
+
+    // Objective function.
+    auto myfunc = [&]( const std::vector<double> x, std::vector<double> grad, void *my_func_data )->double
+    {
+        iter++;
+        auto mu = BS.newParameter();
+        for( int i = 0; i < N; ++i ) mu(i) = x[i];
+        BS.online(mu);
+        auto B = BS.magneticFlux();
+        return B.max();
+    };
+
+    opt.set_max_objective( myfunc, nullptr );
+
+    // inequality constraints
+    auto myconstraint = [&]( const std::vector<double> x, std::vector<double> grad, void *data)->double
+    {
+        auto VT = BS.potentialTemperature();
+        auto T = VT.template element<1>();
+        return T.max() - doption("Tcritic");
+    };
+
+    opt.add_inequality_constraint(myconstraint, nullptr, 1e-8);
+
+    // optimization
+    double minf;
+    tic();
+    std::vector<double> x(N);
+    Eigen::VectorXd::Map(x.data(), N) = muMin;
+
+    ::nlopt::result result = opt.optimize(x, minf);
+
+    double optiTime = toc("optimization", false);
+    Feel::cout << iter << " iterations in " << optiTime << " (" << optiTime/iter << "/iter)" << std::endl;
+
+    // export
+    auto mu = BS.newParameter();
+    for( int i = 0; i < N; ++i ) mu(i) = x[i];
```

Comment:
avoid loops, use copy operator and `Map` from Eigen something like `mu=Map<Eigen::VectorXd>( x.data(), mu.size());`